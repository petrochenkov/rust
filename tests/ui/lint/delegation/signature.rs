#![deny(delegations_detailed, delegations_per_parent_stats)]

use std::hash::{Hash, Hasher};
use std::ops::Add;

trait Foo {
    fn foo(&mut self) -> Self;
    fn bar(&self) -> i32;
}

impl Foo for u32 {
    fn foo(&mut self) -> Self { 1 as u32 }
    fn bar(&self) -> i32 { 1 }
}

struct NonZeroU32(u32);

impl NonZeroU32 {
    fn foo(&mut self) -> Self {
        Self(self.0.foo())
        //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, delegate_to: Other, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: false, caller_has_self: true, same_name: false, ret_postproc: false.
        //~| ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, delegate_to: Field, args_match: Same, args_preproc: false, ret_match: SameUpToSelfType, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: true.
    }

    fn bar(&self) -> i32 {
        self.0.bar()
        //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, delegate_to: Field, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
    }
}

impl Hash for NonZeroU32 {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.0.hash(state);
        //~^ ERROR dstats. parent: TraitImpl, stmts: OneWithoutTail, delegate_to: Field, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
    }
}

impl Add for NonZeroU32 {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Self (self.0.add(other.0))
        //~^ ERROR dstats. parent: TraitImpl, stmts: ZeroWithTail, delegate_to: Other, args_match: Different, args_preproc: false, ret_match: Same, callee_has_self: false, caller_has_self: true, same_name: false, ret_postproc: false.
        //~| ERROR dstats. parent: TraitImpl, stmts: ZeroWithTail, delegate_to: Field, args_match: SameUpToSelfType, args_preproc: true, ret_match: SameUpToSelfType, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: true.
    }
}

mod iter {
    trait Iter {
        type Item;

        fn next(&self) -> Option<Self::Item>;
        fn last(&self) -> Option<Self::Item>;
    }

    impl<T: Iter> Iter for &mut T {
        type Item = T::Item;
        #[inline]
        fn next(&self) -> Option<T::Item> {
            (**self).next()
            //~^ ERROR dstats. parent: TraitImpl, stmts: ZeroWithTail, delegate_to: Other, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
        }
        fn last(&self) -> Option<Self::Item> {
            (**self).last()
            //~^ ERROR dstats. parent: TraitImpl, stmts: ZeroWithTail, delegate_to: Other, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
        }
    }
}

mod anon1 {
    trait Trait {}
    struct Struct;

    fn bar() -> Struct { Struct }

    impl Trait for Struct {}

    impl Struct {
        pub fn new() -> Box<dyn Trait> {
            Box::new(Struct)
            //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, delegate_to: Other, args_match: Different, args_preproc: true, ret_match: Coerced, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
        }

        pub fn bar() -> impl Trait {
            bar()
            //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, delegate_to: Other, args_match: Same, args_preproc: false, ret_match: Same, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
        }
    }
}

mod anon2 {
    use std::fmt::Display;
    struct F<'a>(&'a u32);
    impl<'a> F<'a> {
        fn foo(&self) -> &'a u32 {
            self.0
        }
    }

    fn foo<'a>(f: &F<'a>) -> &'a dyn Display {
        f.foo()
        //~^ ERROR dstats. parent: Other, stmts: ZeroWithTail, delegate_to: Other, args_match: Same, args_preproc: false, ret_match: Coerced, callee_has_self: true, caller_has_self: false, same_name: true, ret_postproc: false.
    }
}

mod coerce {
    use std::fmt::Display;

    struct F;
    impl F {
        fn foo(&self, str: &str) {}
        fn bar(x: &u32) -> &u32 { x }
    }

    struct S(F);

    impl S {
        fn foo(&self, str: &mut str) {
            self.0.foo(str)
            //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, delegate_to: Field, args_match: Coerced, args_preproc: false, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
        }

        fn bar(x: &u32) -> &dyn Display {
            F::bar(x)
            //~^ ERROR dstats. parent: InherentImpl, stmts: ZeroWithTail, delegate_to: Other, args_match: Same, args_preproc: false, ret_match: Coerced, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
        }
    }
}

fn main() {}
