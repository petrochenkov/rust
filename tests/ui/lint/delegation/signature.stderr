error: dstats. parent: Other, stmts: ZeroWithTail, args_match: Same, ret_match: Different, callee_has_self: true, caller_has_self: false, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:101:9
   |
LL | /     fn foo<'a>(f: &F<'a>) -> &'a dyn Display {
LL | |         f.foo()
   | |         ^^^^^^^ callee
LL | |
LL | |     }
   | |_____- caller
   |
note: the lint level is defined here
  --> $DIR/signature.rs:1:9
   |
LL | #![deny(delegations_detailed, delegations_per_parent_stats)]
   |         ^^^^^^^^^^^^^^^^^^^^

error: dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Same, ret_match: Same, callee_has_self: false, caller_has_self: true, same_name: false, ret_postproc: false.
  --> $DIR/signature.rs:20:9
   |
LL | /     fn foo(&mut self) -> Self {
LL | |         Self(self.0.foo())
   | |         ^^^^^^^^^^^^^^^^^^ callee
LL | |
LL | |
LL | |     }
   | |_____- caller

error: dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Same, ret_match: SameUpToSelfType, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: true.
  --> $DIR/signature.rs:20:14
   |
LL | /     fn foo(&mut self) -> Self {
LL | |         Self(self.0.foo())
   | |              ^^^^^^^^^^^^ callee
LL | |
LL | |
LL | |     }
   | |_____- caller

error: dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Same, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:26:9
   |
LL | /     fn bar(&self) -> i32 {
LL | |         self.0.bar()
   | |         ^^^^^^^^^^^^ callee
LL | |
LL | |     }
   | |_____- caller

error: dstats. parent: TraitImpl, stmts: OneWithoutTail, args_match: Same, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:33:9
   |
LL | /     fn hash<H: Hasher>(&self, state: &mut H) {
LL | |         self.0.hash(state);
   | |         ^^^^^^^^^^^^^^^^^^ callee
LL | |
LL | |     }
   | |_____- caller

error: dstats. parent: TraitImpl, stmts: ZeroWithTail, args_match: Different, ret_match: Same, callee_has_self: false, caller_has_self: true, same_name: false, ret_postproc: false.
  --> $DIR/signature.rs:42:9
   |
LL | /     fn add(self, other: Self) -> Self {
LL | |         Self (self.0.add(other.0))
   | |         ^^^^^^^^^^^^^^^^^^^^^^^^^^ callee
LL | |
LL | |
LL | |     }
   | |_____- caller

error: dstats. parent: TraitImpl, stmts: ZeroWithTail, args_match: SameUpToSelfType, ret_match: SameUpToSelfType, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: true.
  --> $DIR/signature.rs:42:15
   |
LL | /     fn add(self, other: Self) -> Self {
LL | |         Self (self.0.add(other.0))
   | |               ^^^^^^^^^^^^^^^^^^^ callee
LL | |
LL | |
LL | |     }
   | |_____- caller

error: dstats. parent: TraitImpl, stmts: ZeroWithTail, args_match: Same, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:60:13
   |
LL | / ...   fn next(&self) -> Option<T::Item> {
LL | | ...       (**self).next()
   | |           ^^^^^^^^^^^^^^^ callee
LL | | ...
LL | | ...   }
   | |_______- caller

error: dstats. parent: TraitImpl, stmts: ZeroWithTail, args_match: Same, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:64:13
   |
LL | / ...   fn last(&self) -> Option<Self::Item> {
LL | | ...       (**self).last()
   | |           ^^^^^^^^^^^^^^^ callee
LL | | ...
LL | | ...   }
   | |_______- caller

error: dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Different, ret_match: Different, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:80:13
   |
LL | / ...   pub fn new() -> Box<dyn Trait> {
LL | | ...       Box::new(Struct)
   | |           ^^^^^^^^^^^^^^^^ callee
LL | | ...
LL | | ...   }
   | |_______- caller

error: dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Same, ret_match: Same, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:85:13
   |
LL | / ...   pub fn bar() -> impl Trait {
LL | | ...       bar()
   | |           ^^^^^ callee
LL | | ...
LL | | ...   }
   | |_______- caller

error: dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: SameNumber, ret_match: Same, callee_has_self: true, caller_has_self: true, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:119:13
   |
LL | / ...   fn foo(&self, str: &mut str) {
LL | | ...       self.0.foo(str)
   | |           ^^^^^^^^^^^^^^^ callee
LL | | ...
LL | | ...   }
   | |_______- caller

error: dstats. parent: InherentImpl, stmts: ZeroWithTail, args_match: Same, ret_match: Different, callee_has_self: false, caller_has_self: false, same_name: true, ret_postproc: false.
  --> $DIR/signature.rs:124:13
   |
LL | / ...   fn bar(x: &u32) -> &dyn Display {
LL | | ...       F::bar(x)
   | |           ^^^^^^^^^ callee
LL | | ...
LL | | ...   }
   | |_______- caller

error: delegation methods stats. methods_count: 1, 5.
   |
note: the lint level is defined here
  --> $DIR/signature.rs:1:31
   |
LL | #![deny(delegations_detailed, delegations_per_parent_stats)]
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: delegation methods stats. methods_count: 2, 2.

error: aborting due to 15 previous errors

